README.md
This repo provides a single `grr` CLI tool to help with using Joyent's code
review (CR) tool, Gerrit, at: <https://cr.joyent.us/>.

Grr is opinionated. The expected workflow is this:

1. You call **`grr <issue>`** to tell grr to create a temporary local branch
   (`grr-<issue>`) for work on this issue (if you are currently on "master"),
   or to use your current branch (if you are already on a non-"master" branch).
   Grr will fetch issue details and remember them (in local git config).
   Optionally you can specify "extra issues" that your changes cover, via
   `grr -a <extra-issue>`. Those issues will be included in the commit message.
2. You make one or more commits, **`git commit -am ...`**, for your change.
3. You call **`grr`** to create/update the CR. Grr will squash the commits (in a
   temporary `grr/auto/<issue>` branch), push the commit (with the appropriate
   commit message) to cr.joyent.us, and remember the CR number.
4. **Get approvals** for your CR, and/or cycle back to step #2.
5. When you get approval, run **`grr`** one last time to update the commit
   message with "Reviewed by" and "Approved by" lines.
6. **Integrate your change** (in the web UI, grr doesn't yet do this), then use
   **`grr -D`** to clean up (delete the `grr-<issue>` branch and switch back to
   master).

See [this example](docs/example.md) for a walk through.


## Install

    npm install -g joyent-grr

Verify that it is installed and on your PATH:

    $ grr --version
    grr 1.2.0
    https://github.com/joyent/grr

You'll need to setup your account in Joyent's Gerrit by following the steps
at https://github.com/joyent/joyent-gerrit/blob/master/docs/user/README.md#new-user-instructions

## Development Hooks

Before commiting be sure to, at least:

    make check      # lint and style checks
    make test       # run tests

A good way to do that is to install the stock pre-commit hook in your
clone via:

    make git-hooks

## License

MPL 2.0


CHANGES.md
# grr changelog

## not yet released

## 1.9.0

- [pull #15] Add support to allow users to grant cr/ia +1s on reviews
  This allows users to pass '-R cr' or '-R ia' to grant CR or IA
  approval on reviews. With the addition of '-d', we print the review
  status of reviews matching the argument, show a diff, and prompt
  the user to grant the approval.

  With '-n', we do a dry-run of granting the approval. Without -d,
  we only list the current approval status of matching reviews.

## 1.8.0

- [issue #12] Add support for multiple issues. (Work mostly by John Levon.)
  This adds the concept of "extra issues" that your CR covers. You can
  specify extra issues initially via:

        grr <main-issue> <extra-issue-1> <extra-issue-2> ...
        grr TRITON-123   TRITON-124 OS-456

  or after you've started via:

        grr -a TRITON-124,OS-456

  You can remove extra issues from a CR via:

        grr -r TRITON-124

  Extra issues will be included in the commit message. E.g. for our
  contrived example:

        TRITON-123 VMAPI metrics test fails in nightly
        TRITON-124 add node-triton support for bhyve
        OS-456 constant dhcp NAKs
        Reviewed-By: ...

## 1.7.0

- [pull #11] Add `grr -L` to list open CRs for the current repository.

## 1.6.0

- Update grr to use the new https://jira.joyent.us

## 1.5.2

- Avoid adding a "Reviewed by: ..." to a CR for a *-1* review.

## 1.5.1

- [pull #2] Get `grr` working with an older git (git 1.7 is Dave's case).

## 1.5.0

- When pushing new commits the author of the latest commit will be used, rather
  than necessarily you. That means you can make a slight tweak to someone
  else's CR, commit your tweak with `git commit --author="The Other Guy
  <guy@example.com>" ...` and then run `grr` without stealing credit for all
  the real work.

## 1.4.0

- Change to using 'grr-$issue' instead of 'grr/$issue' for the created working
  branch, because the '/' might cause issues with Joyent `TRY_BRANCH` builds.

## 1.3.2

- Release 1.3.1 broke things. Fix that.

## 1.3.1

- Allow the origin git URL to not have trailing ".git".

## 1.3.0

- `grr -p PARENTHETICAL` to add a parenthetical to the commit message. Usage:

        $ grr -p "fix make check" ISSUE
        ...
        $ git commit -am "..."
        $ grr

## 1.2.4

- Correct `gerritUsername` in the fix in 1.2.3.


## 1.2.3

**Bad release. Use 1.2.4.**

- Need to use the gerritUsername (that was determined earlier) when doing "ssh
  cr.joyent.us".


## 1.2.2

- Fix `grr` in a branch with a stored JIRA issue. (This was broken in v1.1.0).


## 1.2.1

- Fix `grr -D` assert on `opts.dryRun`.


## 1.2.0

- `grr` will update the commit message to include "Reviewed by" and
  "Approved by" lines according to approvals on the latest patchset.


## 1.1.1

- Fix `AssertionError: opts.dryRun (bool) is required` error for every
  invocation.

## 1.1.0

**Bad release. Use 1.1.1.**

- Support GitHub issues.

- The `<issue>` CLI argument supports a few forms (e.g. the full URL) to simplify
  cutting and pasting:

        grr TOOLS-1531
        grr https://jira.joyent.us/browse/TOOLS-1531
        grr 6
        grr https://github.com/joyent/sdc-triton/issues/6
        grr joyent/sdc-triton#6

- Added commit messages are posted as a review message on the CR.

## 1.0.0

First working version.


lib/common.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2019 Joyent, Inc.
 */

var assert = require('assert-plus');
var forkExecWait = require('forkexec').forkExecWait;
var path = require('path');
var VError = require('verror').VError;


// ---- support stuff

function objCopy(obj, target) {
    assert.object(obj, 'obj');
    assert.optionalObject(obj, 'target');

    if (target === undefined) {
        target = {};
    }
    Object.keys(obj).forEach(function (k) {
        target[k] = obj[k];
    });
    return target;
}


function deepObjCopy(obj) {
    return JSON.parse(JSON.stringify(obj));
}

/**
 * Resolve "~/..." and "~" to an absolute path.
 *
 * Limitations:
 * - This does not handle "~user/...".
 * - This depends on the HOME envvar being defined (%USERPROFILE% on Windows).
 */
function tildeSync(s) {
    var envvar = (process.platform === 'win32' ? 'USERPROFILE' : 'HOME');
    var home = process.env[envvar];
    if (!home) {
        throw new VError('cannot determine home dir: %s environment ' +
            'variable is not defined', envvar);
    }

    if (s === '~') {
        return home;
    } else if (s.slice(0, 2) === '~/' ||
        (process.platform === 'win32' && s.slice(0, 2) === '~'+path.sep))
    {
        return path.resolve(home, s.slice(2));
    } else {
        return s;
    }
}

function indent(s, indentation) {
    if (!indentation) {
        indentation = '    ';
    }
    var lines = s.split(/\r\n|\n|\r/g);
    return indentation + lines.join('\n' + indentation);
}

/*
 * A wrapper around forkExecWait that prints what it is doing.
 *
 * TODO: pass through all forkExecWait opts
 */
function forkExecWaitAndPrint(opts, cb) {
    assert.arrayOfString(opts.argv, 'opts.argv');
    assert.optionalBool(opts.indent, 'opts.indent');
    assert.optionalObject(opts.log, 'opts.log');

    var style = function (s) { return s; };
    if (opts.indent) {
        style = indent;
    }

    // TODO: do better on quoting
    console.log(style('$ ' + opts.argv.join(' ')));

    forkExecWait({argv: opts.argv}, function (err, info) {
        if (opts.log) {
            opts.log.trace({argv: opts.argv, err: err, info: info},
                'forkExecWait');
        }
        if (info.stdout) {
            console.log(style(info.stdout));
        }
        if (info.stderr) {
            console.error(style(info.stderr));
        }
        cb(err, info);
    });
}

function forkExecWaitAndLog(opts, cb) {
    assert.arrayOfString(opts.argv, 'opts.argv');
    assert.object(opts.log, 'opts.log');
    forkExecWait({argv: opts.argv}, function (err, info) {
        opts.log.trace({argv: opts.argv, err: err, info: info}, 'forkExecWait');
        cb(err, info);
    });
}


//---- exports

module.exports = {
    objCopy: objCopy,
    deepObjCopy: deepObjCopy,
    tildeSync: tildeSync,
    indent: indent,
    forkExecWaitAndPrint: forkExecWaitAndPrint,
    forkExecWaitAndLog: forkExecWaitAndLog
};

// vim: set softtabstop=4 shiftwidth=4:


lib/config.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2016 Joyent, Inc.
 */

var assert = require('assert-plus');
var fs = require('fs');
var toml = require('toml');
var VError = require('verror').VError;

var common = require('./common');


var CONFIG_FILE = '~/.grrrc';
var CACHE_FILE = '~/.grrcache';


/**
 * Load the grr config. This is a TOML file at ~/.grrrc
 */
function loadConfigSync(opts) {
    assert.object(opts.log, 'opts.log');

    var config;
    var configPath = common.tildeSync(CONFIG_FILE);
    if (fs.existsSync(configPath)) {
        opts.log.trace({configPath: configPath}, 'loading grr config');
        var raw = fs.readFileSync(configPath, 'utf8');
        try {
            config = toml.parse(raw);
        } catch (e) {
            throw new VError(e, 'invalid config in "%s" at line %s, '
                + 'column %s: %s', CONFIG_FILE, e.line, e.column, e.message);
        }
    } else {
        config = {};
    }

    return config;
}

function loadCacheSync(opts) {
    assert.object(opts.log, 'opts.log');

    var cache;
    var cachePath = common.tildeSync(CACHE_FILE);
    if (fs.existsSync(cachePath)) {
        opts.log.trace({cachePath: cachePath}, 'loading grr cache');
        var raw = fs.readFileSync(cachePath, 'utf8');
        try {
            cache = JSON.parse(raw);
        } catch (err) {
            opts.log.trace({err: err, cachePath: cachePath}, 'invalid cache');
            fs.unlinkSync(cachePath);
            cache = {};
        }
    } else {
        cache = {};
    }

    return cache;
}

function saveCacheSync(opts) {
    assert.object(opts.log, 'opts.log');
    assert.object(opts.cache, 'opts.cache');

    var cachePath = common.tildeSync(CACHE_FILE);
    fs.writeFileSync(cachePath, JSON.stringify(opts.cache));
}



//---- exports

module.exports = {
    CONFIG_FILE: CONFIG_FILE,
    loadConfigSync: loadConfigSync,
    loadCacheSync: loadCacheSync,
    saveCacheSync: saveCacheSync
};

// vim: set softtabstop=4 shiftwidth=4:


lib/cli.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2019 Joyent, Inc.
 */

/*
 * The `grr` CLI
 */

var assert = require('assert-plus');
var bunyan = require('bunyan');
var dashdash = require('dashdash');
var vasync = require('vasync');
var VError = require('verror').VError;

var grr = require('./grr');
var pkg = require('../package.json');


// ---- custom option type (from dashdash/examples/)

function parseCommaSepStringNoEmpties(option, optstr, arg) {
    // JSSTYLED
    return arg.trim().split(/\s*,\s*/g)
        .filter(function (part) { return part; });
}

dashdash.addOptionType({
    name: 'commaSepString',
    takesArg: true,
    helpArg: 'STRING',
    parseArg: parseCommaSepStringNoEmpties
});

dashdash.addOptionType({
    name: 'arrayOfCommaSepString',
    takesArg: true,
    helpArg: 'STRING',
    parseArg: parseCommaSepStringNoEmpties,
    array: true,
    arrayFlatten: true
});


// ---- globals and constants

var log = bunyan.createLogger({
    name: 'grr',
    level: 'warn',
    stream: process.stderr
});

var options = [
    {
        name: 'version',
        type: 'bool',
        help: 'Print version and exit.'
    },
    {
        names: ['help', 'h'],
        type: 'bool',
        help: 'Print this help and exit.'
    },
    {
        names: ['verbose', 'v'],
        type: 'bool',
        help: 'Verbose debugging output. Try `grr -v ... 2>&1 | bunyan`.'
    },
    {
        group: 'For creating/updating CRs'
    },
    {
        names: ['parenthetical', 'p'],
        type: 'string',
        help: 'Add a parenthetical comment to the commit message. Typically '
            + 'used for followup commits on an issue already commited to.',
        helpArg: 'STR'
    },
    {
        names: ['update', 'u'],
        type: 'bool',
        help: 'Re-check cached issue info. This is useful if the issue title '
            + 'was changed.',
        'default': false
    },
    {
        names: ['add-issues', 'a'],
        type: 'arrayOfCommaSepString',
        help: 'Add the given comma-separated list of issues to the commit.',
        helpArg: 'ISSUES'
    },
    {
        names: ['remove-issues', 'r'],
        type: 'arrayOfCommaSepString',
        // JSSTYLED
        help: 'Remove the given comma-separated list of issues from the commit.',
        helpArg: 'ISSUES'
    },
    // TODO
    //{
    //    names: ['commit', 'c'],
    //    type: 'bool',
    //    help: 'Commit local changes.',
    //    'default': false
    //},
    {
        group: 'When done with your CR'
    },
    {
        names: ['delete', 'D'],
        type: 'bool',
        help: 'Delete the local branch used for working with the CR. '
            + 'Typically this is used when you are done with the CR.'
    },
    {
        group: 'Miscellaneous CR operations'
    },
    {
        names: ['list', 'L'],
        type: 'bool',
        help: 'List outstanding code reviews for this repository.'
    },
    {
        names: ['review', 'R'],
        type: 'string',
        help: 'Add either "cr" or "ia" +1s to any matching issues.'
    },
    {   names: ['diff', 'd'],
        type: 'bool',
        help: 'When using -R, show the latest diffs of reviews.'
    },
    {
        names: ['dry-run', 'n'],
        type: 'bool',
        help: 'Go through the motions without actually committing, pushing or '
            + 'reviewing.',
        'default': false
    }
];

var parser = dashdash.createParser({options: options});

var VALID_REVIEWS = {
    'cr': 'Code-Review',
    'ia': 'Integration-Approval'};

// ---- support functions

function usage() {
    var help = parser.help({includeEnv: true}).trimRight();
    /* BEGIN JSSTYLED */
    console.log([
        'Opinionated hand-holding of code review (CR) mechanics in Joyent\'s',
        'Gerrit at <https://cr.joyent.us>.',
        '',
        'Usage:',
        '    grr [<options>] [<issue> [<extra-issues>]]',
        '',
        '      (no options)          # start or update a CR',
        '      -D                    # delete feature branch, switch to master',
        '      -L                    # list outstanding CRs for this repo',
        '      -R <ia|cr> [-d] [-n] <issue>',
        '                            # grant or list +1s for matching issues',
        '',
        'Options:',
        help,
        '',
        'Where <issue> is a JIRA issue (the PROJ-123 name or the full URL) or',
        'a GitHub issue (the number, account/repo#num, or full URL).',
        '',
        'Examples:',
        '    Working on changes:',
        '',
        '    grr TOOLS-123    # Start or update a branch and CR for this issue',
        //'    grr --commit TOOLS-123    # ... also commit uncommitted changes',
        '    grr              # Update the current CR (using the cached issue)',
        '    grr -a TOOLS-124 # Add an extra issue to the current CR',
        '    grr -r TOOLS-124 # Remove an extra issue from the current CR',
        '    grr -D           # All done. Delete the branch and switch',
        '                     # back to master',
        '    grr -L           # Show reviews in progress for this repository,',
        '                     # useful to see whether your changes might clash',
        '                     # with other work.',
        '',
        '    Reviewing changes:',
        '',
        '    grr -R cr TOOLS-124       # show review status for TOOLS-124',
        '    grr -R cr -d -n TOOLS-124 # show diffs for TOOLS-124',
        '                              # but only do a dry-run review.',
        '    grr -R ia -d TOOLS-124    # show diffs for TOOLS-124,',
        '                              # and prompt for IA approval for each.'
    ].join('\n'));
    /* END JSSTYLED */
}


function mainFinish(err) {
    if (err) {
        var exitStatus = err.exitStatus || 1;
        var showErrStack = (log.level() <= bunyan.DEBUG); // use --verbose
        console.error('grr: error: %s',
            (showErrStack ? err.stack : err.message));

        // Use a soft exit (i.e. just set `process.exitCode`) if supported.
        var supportsProcessExitCode = true;
        var nodeVer = process.versions.node.split('.').map(Number);
        if (nodeVer[0] === 0 && nodeVer[1] <= 10) {
            supportsProcessExitCode = false;
        }
        if (supportsProcessExitCode) {
            process.exitCode = exitStatus;
        } else if (exitStatus !== 0) {
            process.exit(exitStatus);
        }
    }
}

function fatal(errmsg) {
    console.error('grr: error: %s', errmsg);
    process.exit(1);
}


// ---- mainline

function main(argv) {
    // 'actions' is plural because coming "submit" work might allow multiple.
    var actions = [];
    var issueArg;
    var addExtraIssueArgs;
    var removeExtraIssueArgs;
    var opts;

    try {
        opts = parser.parse(process.argv);
    } catch (e) {
        fatal(e.message);
    }

    if (opts.help) {
        usage();
        process.exit(0);
    } else if (opts.version) {
        console.log('grr ' + pkg.version);
        console.log(pkg.homepage);
        process.exit(0);
    }

    if (opts.verbose) {
        log.level('trace');
        log.src = true;
    }

    if ((opts._args.length > 0 || opts.update || opts.add_issues ||
        opts.remove_issues) && !(opts.list || opts.review))
    {
        actions.push('update-or-create');
        if (opts._args.length > 0) {
            issueArg = opts._args.shift();
        }
        addExtraIssueArgs = opts._args.slice();
        if (opts.add_issues) {
            addExtraIssueArgs = addExtraIssueArgs.concat(opts.add_issues);
        }
        if (opts.remove_issues) {
            removeExtraIssueArgs = opts.remove_issues;
        }
    }

    if (opts.delete) {
        if (actions.length) {
            fatal('cannot delete (-D) when reviewing (-R), listing (-L), or '
                + 'creating/updating a CR');
        }
        actions.push('delete');
    }

    if (opts.list) {
        if (actions.length) {
            fatal('cannot list (-L) when deleting (-D), reviewing (-R) or '
                + 'creating/updating a CR');
        }
        actions.push('list');
    }

    if (opts.review) {
        if (actions.length) {
            fatal('cannot review (-R) when deleting (-D), listing (-L) or '
                + 'creating/updating a CR');
        }
        if (!VALID_REVIEWS.hasOwnProperty(opts.review)) {
            fatal('unknown review type ' + opts.review);
        }
        actions.push('review');
        if (opts._args.length === 1) {
            issueArg = opts._args.shift();
        } else {
            fatal('single issue required when using -R');
        }
    }

    if (actions.length === 0) {
        actions.push('update-or-create');
    }

    // Only one action at a time for now.
    assert.equal(actions.length, 1,
        'incorrect number of actions: ' + JSON.stringify(actions));

    switch (actions[0]) {
        case 'list':
            grr.grrList({log: log}, mainFinish);
            break;
        case 'delete':
            grr.grrDelete({
                log: log,
                dryRun: opts.dry_run
            }, mainFinish);
            break;
        case 'update-or-create':
            grr.grrUpdateOrCreate({
                log: log,
                issueArg: issueArg,
                parenthetical: opts.parenthetical,
                addExtraIssueArgs: addExtraIssueArgs,
                removeExtraIssueArgs: removeExtraIssueArgs,
                dryRun: opts.dry_run,
                commit: opts.commit,
                update: opts.update
            }, mainFinish);
            break;
        case 'review':
            grr.grrReview({
                log: log,
                issueArg: issueArg,
                dryRun: opts.dry_run,
                reviewArg: VALID_REVIEWS[opts.review],
                diffArg: opts.diff,
                verbose: opts.verbose}, mainFinish);
            break;
        default:
            fatal('what action is this? ' + actions[0]);
            break;
    }
}


// ---- exports

module.exports = {
    main: main
};

// vim: set softtabstop=4 shiftwidth=4:
